### PR TITLE
Updated mosaic restrictions endpoints and DTOs

### PR DESCRIPTION
**Pull request by martinayora**

### CODE REVIEW

**Comment by martinayora** [comment]
As far as I know openApi specification isn't supporting duplicated paths, even with different method. I'm getting a conflict here with the _**/restrictions/mosaic/{mosaicId}**_ endpoint.

Any suggestions on how to fix this? @dgarcia360 @Vektrat

Documentation shouldn't affect development. But maybe it is a wrong REST approach and we should change endpoint names.

First _**/restrictions/mosaic/{mosaicId}**_ is retrieving _mosaic global restrictions_. And the second one _mosaic address restrictions_.

**Comment by Vektrat** [comment]
> As far as I know openApi specification isn't supporting duplicated paths, even with different method. I'm getting a conflict here with the _**/restrictions/mosaic/{mosaicId}**_ endpoint.

the same endpoint for multiple methods is extremely common and valid, there has to be a way for a single endpoint to allow multiple methods

**Comment by dgarcia360** [comment]
@Vektrat @martinayora Fixed the errors raised in this comment, you can merge different verbs under the same endpoint path: https://github.com/nemtech/nem2-openapi/pull/29/commits/9cac8582b798b04080da07a32b821cade8431aca

Let me know if this is what you'd like to achieve.

**Comment by martinayora** [comment]
Well, that was simple. 🤣  Perfect then. 👍

**Comment by dgarcia360** [comment]
@rg911 Could you add the comment in the lines affected? Thanks!

File: spec/openapi3.yaml [hunk 1]
```diff
@@ -3165,6 +3253,87 @@ components:
...
+      type: object
+      required:
+        - compositeHash
+        - entryType
+        - mosaicId
+        - restrictions
+      properties:
+        compositeHash:
+          $ref: "#/components/schemas/Hash256"
+        entryType:
```
Comment (rg911) [inline spec/openapi3.yaml]: So this is actually the restrictionType enum 0: address, 1: global. Should a enum be created for this?

File: spec/openapi3.yaml [hunk 2]
```diff
@@ -3165,6 +3253,87 @@ components:
...
+        - mosaicId
+        - targetAddress
+        - restrictions
+      properties:
+        compositeHash:
+          $ref: "#/components/schemas/Hash256"
+        entryType:
+          type: integer
+        mosaicId:
+          $ref: "#/components/schemas/MosaicId"
```
[the remark this answers is not in the document]
Comment (dgarcia360) [reply spec/openapi3.yaml]: Why not:

    MosaicId:
      type: string
      format: hex
      description: Mosaic identifier.
      example: 0DC67FBE1CAD29E3

File: spec/openapi3.yaml [hunk 3]
```diff
@@ -3165,6 +3253,87 @@ components:
...
+        - entryType
+        - mosaicId
+        - restrictions
+      properties:
+        compositeHash:
+          $ref: "#/components/schemas/Hash256"
+        entryType:
+          type: integer
+        mosaicId:
+          $ref: "#/components/schemas/MosaicId"
```
[the remark this answers is not in the document]
Comment (dgarcia360) [reply spec/openapi3.yaml]: Whe not? 

    MosaicId:
      type: string
      format: hex
      description: Mosaic identifier.
      example: 0DC67FBE1CAD29E3

File: spec/openapi3.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (rg911) [reply spec/openapi3.yaml]: even better 👍

[the remark this answers is not in the document]
Comment (dgarcia360) [reply spec/openapi3.yaml]: $ref: "#/components/schemas/MosaicId" is linking to the definition previously shared.

**Comment by martinayora** [comment]
> mosaicRestriction DTO models are missing `mosaicRestrictionEntry` wrapper

It is on top of each of the DTOs.  But I may have left some.

File: spec/openapi3.yaml [hunk 4]
```diff
@@ -3165,6 +3263,101 @@ components:
...
+      required:
+        - key
+        - restriction
+      properties:
+        key:
+          type: string
+          description: Restriction key.
+          example: 0DC67FBE1CAD29E3
+        restriction:
+          type: array
```
Comment (rg911) [inline spec/openapi3.yaml]: by comparing the payload from the server, this I believe should be an `object`

File: spec/openapi3.yaml [hunk 5]
```diff
@@ -3165,6 +3263,99 @@ components:
...
+        - compositeHash
+        - entryType
+        - mosaicId
+        - targetAddress
+        - restrictions
+      properties:
+        compositeHash:
+          $ref: "#/components/schemas/Hash256"
+        entryType:
+          type: integer
```
Comment (dgarcia360) [inline spec/openapi3.yaml]: $ref: "#/components/schemas/MosaicRestrictionEntryTypeEnum"

File: spec/openapi3.yaml [hunk 6]
```diff
@@ -3165,6 +3263,99 @@ components:
...
+        - restrictionValue
+        - restrictionType
+      properties:
+        referenceMosaicId:
+          $ref: "#/components/schemas/MosaicId"
+        restrictionValue:
+          type: string
+          description: Restriction value.
+        restrictionType:
+          type: integer
```
Comment (dgarcia360) [inline spec/openapi3.yaml]: $ref: "#/components/schemas/MosaicRestrictionTypeEnum"

File: spec/openapi3.yaml [hunk 7]
```diff
@@ -3165,6 +3263,99 @@ components:
...
+    MosaicGlobalRestrictionEntryDTO:
+      type: object
+      required:
+        - key
+        - restriction
+      properties:
+        key:
+          type: string
+          description: Restriction key.
```
Comment (dgarcia360) [inline spec/openapi3.yaml]: format: hex

File: spec/openapi3.yaml [hunk 8]
```diff
@@ -3165,6 +3263,99 @@ components:
...
+          $ref: "#/components/schemas/MosaicGlobalRestrictionEntryWrapperDTO"
+    MosaicAddressRestrictionEntryDTO:
+      type: object
+      required:
+        - key
+        - value
+      properties:
+        key:
+          type: string
+          description: Restriction key.
```
Comment (dgarcia360) [inline spec/openapi3.yaml]: format: hex

Comment (dgarcia360) [reply spec/openapi3.yaml]: We can also abstract the MosaicRestrictionKey as a separate object as we are using it in multiple places.

**Pull request merged**